sim/shaman/enhancement/rotation.go

```diff
@@ -78,46 +54,38 @@ func (rotation *AdaptiveRotation) DoAction(enh *EnhancementShaman, sim *core.Sim
 			timeUntilSwing = enh.AutoAttacks.MH.SwingDuration
 		}
 		if lbCastTime < timeUntilSwing {
-			if !enh.LightningBolt.Cast(sim, target) {
-				enh.WaitForMana(sim, enh.LightningBolt.CurCast.Cost)
-			}
-			return
+			return enh.LightningBolt
 		}
 	}
-
 	if enh.EarthShock.IsReady(sim) {
-		if !enh.EarthShock.Cast(sim, target) {
-			enh.WaitForMana(sim, enh.EarthShock.CurCast.Cost)
-		}
-		return
+		return enh.EarthShock
 	}
-
-	if !enh.LightningShieldAura.IsActive() {
-		enh.LightningShield.Cast(sim, nil)
-		return
+	if !enh.LightningShieldAura.IsActive() && enh.LightningShield.IsReady(sim) {
+		return enh.LightningShield
 	}
-
 	if enh.Totems.Fire != proto.FireTotem_NoFireTotem {
 		if enh.FireNova.IsReady(sim) && enh.CurrentMana() > 4000 {
-			if !enh.FireNova.Cast(sim, target) {
-				enh.WaitForMana(sim, enh.FireNova.CurCast.Cost)
-			}
-			return
+			return enh.FireNova
 		}
 	}
+	if enh.Talents.LavaLash && enh.LavaLash.IsReady(sim) {
+		return enh.LavaLash
+	}
+	return nil
+}
 
-	if enh.Talents.LavaLash {
-		if enh.LavaLash.IsReady(sim) {
-			if !enh.LavaLash.Cast(sim, target) {
-				enh.WaitForMana(sim, enh.LavaLash.CurCast.Cost)
-			}
-			return
-		}
+func (rotation *AdaptiveRotation) DoAction(enh *EnhancementShaman, sim *core.Simulation) {
+	target := sim.GetTargetUnit(0)
+
+	spell := rotation.chooseSpell(enh, sim, target)
+	if spell == nil {
+		enh.DoNothing()
+		return
 	}
 
-	//enh.LightningShield.Cast(sim, nil) // if nothing else, refresh lightning shield
-	enh.DoNothing()
-	return
+	if !spell.Cast(sim, target) {
+		enh.WaitForMana(sim, spell.CurCast.Cost)
+	}
 }
```

Comment:
Looks alot better, but one more thing I can think of can be done before merging.

I _really_ do not like the `isReady()`-check within the `chooseSpell`. I haven't checked, maybe you have. Is it possible to move it to `DoAction`?

I recognize something along the lines of, but I am not sure if that changes the landscape of things.

```go
    spell := rotation.chooseSpell(enh, sim, target)
	if spell.IsReady() && !spell.Cast(sim, target) {
		enh.WaitForMana(sim, spell.CurCast.Cost)
	}
```

Reply:
As per discussion on discord - If we will use `IsReady` outside of the `chooseSpell` then we need to have an option to fallback to it. This means solving this in a loop or preparing and checking chooseSpell only on Ready spells. Let's do one of those in next iteration